File: python/hamming.py
def generate_hamming_code(data_bits: str) -> str:
    """
    Generate a Hamming(7,4) code from 4 data bits.

    Args:
        data_bits (str): A string of 4 binary digits.

    Returns:
        str: A string of 7 binary digits representing the Hamming code.
    """
    data = list(map(int, data_bits))
    m = len(data)
    r = 3  # For Hamming(7,4), we have 3 parity bits

    hamming_code = [0] * (m + r)
    j = 0
    for i in range(1, len(hamming_code) + 1):
        if (i & (i - 1)) == 0:
            continue
        hamming_code[i - 1] = data[j]
        j += 1

    # Calculate parity bits
    for i in range(r):
        parity_pos = 2 ** i
        parity = 0
        for j in range(parity_pos, len(hamming_code) + 1):
            if j & parity_pos:
                parity ^= hamming_code[j - 1]
        hamming_code[parity_pos - 1] = parity

    return ''.join(map(str, hamming_code))

def detect_and_correct_error(hamming_code: str) -> str:
    """
    Detect and correct a single-bit error in a Hamming(7,4) code.

    Args:
        hamming_code (str): A string of 7 binary digits.

    Returns:
        str: A string of 4 binary digits representing the corrected data.
    """
    hamming = list(map(int, hamming_code))
    n = len(hamming)
    r = 3  # For Hamming(7,4), we have 3 parity bits

    error_pos = 0
    for i in range(r):
        parity_pos = 2 ** i
        parity = 0
        for j in range(parity_pos, n + 1):
            if j & parity_pos:
                parity ^= hamming[j - 1]
        if parity != 0:
            error_pos += parity_pos

    if error_pos != 0:
        hamming[error_pos - 1] ^= 1

    data_bits = []
    for i in range(1, n + 1):
        if (i & (i - 1)) != 0:
            data_bits.append(hamming[i - 1])

    return ''.join(map(str, data_bits))

def encode_binary_string(data_bits: str) -> str:
    """
    Encode a binary string using Hamming(7,4) encoding.

    Args:
        data_bits (str): A binary string to encode.

    Returns:
        str: The encoded binary string with padding information.
    """
    padding_length = (4 - len(data_bits) % 4) % 4
    padded_data_bits = data_bits + '0' * padding_length

    encoded_parts = []
    for i in range(0, len(padded_data_bits), 4):
        part = padded_data_bits[i:i+4]
        hamming_code = generate_hamming_code(part)
        encoded_parts.append(hamming_code)

    # Append the padding length as a 4-bit binary string at the end
    padding_info = format(padding_length, '03b') + '1'
    binary_string = ''.join(encoded_parts) + padding_info

    # Calculate additional padding to make the length a multiple of 8
    total_padding_length = (8 - len(binary_string) % 8) % 8
    binary_string += '0' * total_padding_length

    return binary_string

def decode_binary_string(encoded_string: str) -> str:
    """
    Decode a binary string encoded with Hamming(7,4) and correct errors.

    Args:
        encoded_string (str): The encoded binary string with padding information.

    Returns:
        str: The original binary string after decoding and error correction.
    """
    # Find the rightmost '1' to identify the padding info
    padding_info_index = encoded_string.rfind('1')
    padding_info = encoded_string[padding_info_index-3:padding_info_index]
    padding_length = int(padding_info, 2)

    # Remove the padding info from the encoded string
    encoded_string = encoded_string[:padding_info_index-3]

    decoded_parts = []
    for i in range(0, len(encoded_string), 7):
        hamming_code = encoded_string[i:i+7]
        corrected_code = detect_and_correct_error(hamming_code)
        decoded_parts.append(corrected_code)

    decoded_binary_string = ''.join(decoded_parts)

    # Remove the padding that was added earlier
    if padding_length > 0:
        decoded_binary_string = decoded_binary_string[:-padding_length]

    return decoded_binary_string

def bytes_to_binary_string(byte_data: bytes) -> str:
    """
    Convert bytes to an 8-bit binary string.

    Args:
        byte_data (bytes): The byte data to convert.

    Returns:
        str: The binary string representation of the byte data.
    """
    return ''.join(format(byte, '08b') for byte in byte_data)

def binary_string_to_bytes(binary_str: str) -> bytes:
    """
    Convert a binary string back to bytes.

    Args:
        binary_str (str): The binary string to convert. Must be a multiple of 8.

    Returns:
        bytes: The byte representation of the binary string.
    """
    if len(binary_str) % 8 != 0:
        raise ValueError("Binary string length must be a multiple of 8")

    return bytes(int(binary_str[i:i+8], 2) for i in range(0, len(binary_str), 8))

# Example usage
if __name__ == "__main__":
    # Convert "Hello world" to a binary string
    message = "Hello world!"
    binary_string = ''.join(format(ord(char), '08b') for char in message)
    print(f'Original binary string: {binary_string}')

    # Encode the binary string
    encoded_string = encode_binary_string(binary_string)
    print(f'Encoded binary string: {encoded_string}')

    # Introduce an error in the encoded string
    encoded_string_with_error = list(encoded_string)
    if encoded_string_with_error[16] == '0':
        encoded_string_with_error[16] = '1'
    else:
        encoded_string_with_error[16] = '0'
    encoded_string_with_error = ''.join(encoded_string_with_error)
    print(f'Encoded string with error: {encoded_string_with_error}')

    # Decode and correct the error
    corrected_string = decode_binary_string(encoded_string_with_error)
    print(f'Corrected binary string: {corrected_string}')

    # Convert the corrected binary string back to text
    decoded_text = ''.join(chr(int(corrected_string[i:i+8], 2)) for i in range(0, len(corrected_string), 8))
    print(f'Decoded text: {decoded_text}')
    print("Original message: ", message)
    print("Length of original message: ", len(message))
    print("Length of decoded message: ", len(decoded_text))

    # Show different chars
    print("Different chars: ", set(decoded_text) - set(message))

    assert decoded_text == message

File: C/hamming.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

// Function to convert a character to an 8-bit binary string
void char_to_binary(char c, char *binary) {
    for (int i = 7; i >= 0; i--) {
        binary[7 - i] = ((c >> i) & 1) + '0';
    }
    binary[8] = '\0';
}

// Function to generate Hamming(7,4) code
void generate_hamming_code(const char *data_bits, char *hamming_code) {
    int data[4];
    for (int i = 0; i < 4; i++) {
        data[i] = data_bits[i] - '0';
    }

    // Calculate parity bits
    int p1 = data[0] ^ data[1] ^ data[3];
    int p2 = data[0] ^ data[2] ^ data[3];
    int p3 = data[1] ^ data[2] ^ data[3];

    // Construct the Hamming code
    hamming_code[0] = p1 + '0';
    hamming_code[1] = p2 + '0';
    hamming_code[2] = data[0] + '0';
    hamming_code[3] = p3 + '0';
    hamming_code[4] = data[1] + '0';
    hamming_code[5] = data[2] + '0';
    hamming_code[6] = data[3] + '0';
    hamming_code[7] = '\0';
}

// Function to detect and correct a single-bit error in a Hamming(7,4) code
void detect_and_correct_error(char *hamming_code) {
    int hamming[7];
    for (int i = 0; i < 7; i++) {
        hamming[i] = hamming_code[i] - '0';
    }

    // Calculate parity checks
    int p1 = hamming[0] ^ hamming[2] ^ hamming[4] ^ hamming[6];
    int p2 = hamming[1] ^ hamming[2] ^ hamming[5] ^ hamming[6];
    int p3 = hamming[3] ^ hamming[4] ^ hamming[5] ^ hamming[6];

    // Calculate error position
    int error_pos = p1 * 1 + p2 * 2 + p3 * 4;

    // Correct the error if there is one
    if (error_pos != 0) {
        hamming[error_pos - 1] ^= 1;
    }

    // Update the corrected Hamming code
    for (int i = 0; i < 7; i++) {
        hamming_code[i] = hamming[i] + '0';
    }
}

// Function to encode a binary string using Hamming(7,4) encoding
void encode_binary_string(const char *data_bits, char *encoded_string) {
    size_t len = strlen(data_bits);
    char hamming_code[8];
    char *encoded_ptr = encoded_string;

    for (size_t i = 0; i < len; i += 4) {
        generate_hamming_code(data_bits + i, hamming_code);
        memcpy(encoded_ptr, hamming_code, 7);
        encoded_ptr += 7;
    }
    *encoded_ptr = '\0';
}

// Function to decode a binary string encoded with Hamming(7,4) and correct errors
void decode_binary_string(const char *encoded_string, char *decoded_string) {
    size_t len = strlen(encoded_string);
    char hamming_code[8];
    char corrected_code[5];
    char *decoded_ptr = decoded_string;

    for (size_t i = 0; i < len; i += 7) {
        strncpy(hamming_code, encoded_string + i, 7);
        hamming_code[7] = '\0';
        detect_and_correct_error(hamming_code);

        // Extract the original data bits
        corrected_code[0] = hamming_code[2];
        corrected_code[1] = hamming_code[4];
        corrected_code[2] = hamming_code[5];
        corrected_code[3] = hamming_code[6];
        memcpy(decoded_ptr, corrected_code, 4);
        decoded_ptr += 4;
    }
    *decoded_ptr = '\0';
}

// Main function for CLI tool
int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <encode|decode> [data]\n", argv[0]);
        return 1;
    }

    char *operation = argv[1];
    char *input_string = NULL;
    size_t buffer_size = 1024;
    size_t total_read = 0;
    size_t bytes_read;

    // Allocate initial buffer
    input_string = malloc(buffer_size);
    if (!input_string) {
        fprintf(stderr, "Memory allocation failed\n");
        return 1;
    }

    // Read from stdin if data is available
    while ((bytes_read = fread(input_string + total_read, 1, buffer_size - total_read, stdin)) > 0) {
        total_read += bytes_read;
        // Reallocate buffer if necessary
        if (total_read == buffer_size) {
            buffer_size *= 2;
            input_string = realloc(input_string, buffer_size);
            if (!input_string) {
                fprintf(stderr, "Memory reallocation failed\n");
                return 1;
            }
        }
    }

    // Null-terminate the string
    input_string[total_read] = '\0';

    if (total_read == 0 && argc == 3) {
        // Use command-line argument if no stdin input
        input_string = argv[2];
    } else if (total_read == 0) {
        fprintf(stderr, "No input data provided.\n");
        free(input_string);
        return 1;
    }

    if (strcmp(operation, "encode") == 0) {
        // Convert the input string to a binary string
        size_t input_len = strlen(input_string);
        char *binary_string = (char *)malloc(input_len * 8 + 1);
        if (!binary_string) {
            fprintf(stderr, "Memory allocation failed\n");
            free(input_string);
            return 1;
        }
        char *binary_ptr = binary_string;
        for (size_t i = 0; i < input_len; i++) {
            char binary_char[9];
            char_to_binary(input_string[i], binary_char);
            memcpy(binary_ptr, binary_char, 8);
            binary_ptr += 8;
        }
        *binary_ptr = '\0';

        // Encode the binary string
        char *encoded_string = (char *)malloc(input_len * 14 + 1);  // Rough estimate for encoded size
        if (!encoded_string) {
            fprintf(stderr, "Memory allocation failed\n");
            free(binary_string);
            free(input_string);
            return 1;
        }
        encode_binary_string(binary_string, encoded_string);
        printf("%s", encoded_string);

        free(binary_string);
        free(encoded_string);
    } else if (strcmp(operation, "decode") == 0) {
        // Decode the binary string
        size_t input_len = strlen(input_string);
        char *corrected_string = (char *)malloc(input_len / 7 * 4 + 1);
        if (!corrected_string) {
            fprintf(stderr, "Memory allocation failed\n");
            free(input_string);
            return 1;
        }
        decode_binary_string(input_string, corrected_string);

        // Convert the corrected binary string back to text
        size_t corrected_len = strlen(corrected_string);
        char *decoded_text = (char *)malloc(corrected_len / 8 + 1);
        if (!decoded_text) {
            fprintf(stderr, "Memory allocation failed\n");
            free(corrected_string);
            free(input_string);
            return 1;
        }
        char *decoded_ptr = decoded_text;
        for (size_t i = 0; i < corrected_len; i += 8) {
            char byte_str[9];
            strncpy(byte_str, corrected_string + i, 8);
            byte_str[8] = '\0';
            *decoded_ptr++ = (char)strtol(byte_str, NULL, 2);
        }
        *decoded_ptr = '\0';
        printf("%s", decoded_text);

        free(corrected_string);
        free(decoded_text);
    } else {
        fprintf(stderr, "Invalid operation. Use 'encode' or 'decode'.\n");
        free(input_string);
        return 1;
    }

    free(input_string);
    return 0;
}